File: spring/member-server/src/main/java/com/example/memberserver/member/MemberCommandService.java
package com.example.memberserver.member;

import com.example.memberserver.common.validator.ValidatorTemplate;
import com.example.memberserver.infrastructure.IdProvier.IdpClient;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor
@Service
public class MemberCommandService {
    private final IdpClient idpClient;
    private final MemberRepository memberRepository;
    private final ValidatorTemplate<MemberDto> memberValidator;

    @Transactional(rollbackFor = Exception.class)
    public Long create(MemberDto memberDto) {

        memberValidator.execute(memberDto);

        Member member = Member.of(memberDto);
        Member savedMember = memberRepository.save(member);
        return idpClient.createMember(savedMember);
    }

    @Transactional(rollbackFor = Exception.class)
    public Long update(Long memberId, MemberDto memberDto) {

        memberValidator.execute(memberDto);

        Member member = memberRepository.findById(memberId).orElseThrow(() -> new RuntimeException("Member not found"));
        Member updatedMember = member.update(memberDto);
        memberRepository.save(updatedMember);
        return idpClient.updateMember(updatedMember);
    }

    @Transactional(rollbackFor = Exception.class)
    public void delete(Long memberId) {
        Member member = memberRepository.findById(memberId).orElseThrow(() -> new RuntimeException("Member not found"));
        memberRepository.delete(member);
        idpClient.deleteMember(member);
    }
}

File: user-server-firebase-spring-kotlin/src/main/kotlin/com/example/userserverfirebasespringkotlin/MemberCommandService.kt
package com.example.userserverfirebasespringkotlin

import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class MemberCommandService(
    private val idpClient: IdpClient,
    private val memberRepository: MemberRepository,
    private val permissionRepository: PermissionRepository,
) {

    @Transactional(rollbackFor = [Exception::class])
    fun create(memberDto: MemberDto): Long? {
        if (memberDto.permissionIds.isNotEmpty() || memberDto.password.isEmpty() || memberDto.email.isEmpty()) {
            throw Exception()
        }
        val member = Member.of(memberDto)
        val savedMember = memberRepository.save(member)
        return idpClient.createMember(member = savedMember, password = memberDto.password)
    }

    @Transactional(rollbackFor = [Exception::class])
    fun update(memberId: Long, memberDto: MemberDto): Long? {
        val member = memberRepository.findByIdOrNull(memberId) ?: throw Exception()
        val permissions = permissionRepository.findAllByIdIn(memberDto.permissionIds.map { it.toLong() })
        val updatedMember = member.assignPermissions(permissions)
        memberRepository.save(updatedMember)
        return idpClient.updateMember(updatedMember)
    }

    @Transactional(rollbackFor = [Exception::class])
    fun delete(memberId: Long) {
        val member = memberRepository.findByIdOrNull(memberId) ?: throw Exception()
        memberRepository.delete(member)
        idpClient.deleteMember(member)
    }

}
